monai/networks/nets/milmodel.py

```diff
@@ -0,0 +1,184 @@
+import os
+
+import torch
+import torch.nn as nn
+import torchvision.models as models
+
+
+class MILModel(nn.Module):
+    """
+    A wrapper around backbone classification model suitable for MIL
+
+    Args:
+        num_classes: number of output classes
+        mil_mode: MIL variant (supported max, mean, att, att_trans, att_trans_pyramid
+        pretrained: init backbone with pretrained weights. Defaults to True.
+        backbone: Backbone classifier CNN. Defaults to None, it which case ResNet50 will be used.
+        backbone_nfeatures: Number of output featues of the backbone CNN (necessary only when using custom backbone)
+
+    mil_mode:
+        mean - average features from all instances, equivalent to pure CNN (non MIL)
+        max - retain only the instance with the max probability for loss calculation
+        att - attention based MIL https://arxiv.org/abs/1802.04712
+        att_trans - transformer MIL
+        att_trans_pyramid - transformer pyramid MIL
+
+    """
+
+    def __init__(self, num_classes, mil_mode="att", pretrained=True, backbone=None, backbone_nfeatures=None):
+
+        super().__init__()
+
+        self.mil_mode = mil_mode
+        print("MILModel with mode", mil_mode, "num_classes", num_classes)
+        n_trans = 4
+        trans_dropout = 0.0
+        self.attention = None
+
+        if backbone is None:
+            # use ResNet50 backbone
+
+            # os.environ["TORCH_HOME"] = "../../torchhome"
+            # import torchvision.models as models
+
+            net = models.resnet50(pretrained=pretrained)
+            nfc = net.fc.in_features  # save the number of final features
+            net.fc = torch.nn.Identity()  # remove final linear layer
+
+            self.extra_outputs = {}
+            if mil_mode == "att_trans_pyramid":
+                # register hooks to capture outputs of intermediate layers
+                def forward_hook(layer_name):
+                    def hook(module, input, output):
+                        self.extra_outputs[layer_name] = output
+
+                    return hook
+
+                self.fhooks = []
+                net.layer1.register_forward_hook(forward_hook("layer1"))
+                net.layer2.register_forward_hook(forward_hook("layer2"))
+                net.layer3.register_forward_hook(forward_hook("layer3"))
+                net.layer4.register_forward_hook(forward_hook("layer4"))
+
+        else:
+            # use a custom backbone (untested)
+            net = backbone
```

Comment:
It would be nice that this class can accept the name of the torchvision model as well., like `"densenet161"`.
So it can check if the backbone is `str`, it imports the model like this:
https://github.com/Project-MONAI/MONAI/blob/3960a514a5f44a10aceab463e926a19bc0b8c79a/monai/networks/nets/torchvision_fc.py#L61

Reply:
hmm, ok

Reply:
but we also need to supply the dimension of features space, as well as we need to remove the last FC layer, it'll be hard to do an arbitrary torchvision model based on string alone

Reply:
Right, but you are already accepting any arbitrary model that is derived from `nn.Module`. Also how is it different than `net = models.resnet50(pretrained=pretrained)`?

Reply:
if someone wants to provide ResNet50 as an arbitrary model , they will have to do these steps:
net = models.resnet50(pretrained=pretrained) #init
net.fc = nn.Identity() #remove final FC
nfc = 2048 # define output features size

model = MILModel(backbone=net, backbone_nfeatures=nfc)

the final FC layer for ResNet50 is called .fc , but for other networks it can be called different, so for arbitrary network we don't know how to remove it

Reply:
@myron, I see what you are saying but it is not very different than `netadapter`. The FC layer does not need to be replaces by identity, it can be removed like here: https://github.com/Project-MONAI/MONAI/blob/3960a514a5f44a10aceab463e926a19bc0b8c79a/monai/networks/nets/netadapter.py#L75

And this can work for a large family of torchvision models.

Reply:
this solution is not generic. the final layer of torchvision model may or may not be FC layer , it depends on when FC layer was initialised. Furthermore, the final layer may be nn.Sequential which will also produce incorrect results.  Using this approach will hide many potential errors , I would avoid using it